### PR TITLE
[stable/kong] refactor env block generation

### PR DESCRIPTION
**Pull request by rainest**
#### What this PR does / why we need it:
tl;dr it makes generating EnvVar blocks cleaner and allows us to guarantee that we only provide one version of an environment variable, because providing multiple versions causes bad things to happen.

This change significantly refactors the generation of Kong container env blocks. It also consolidates various secretKeyRef templates into a single generic template, and makes several minor changes elsewhere in the chart to accomodate the new templates.

The current template handles user-provided environment variables and automatically-generated variables differently. User variables are rendered in a loop over the `env` section of values.yaml, and generated variable values are inserted into a static template. If a user provides a variable that is also generated automatically, the resulting EnvVar block will contain both versions of the variable.

The new template adds both user and generated variables to dictionaries, which are then combined and run through a render loop to generate the EnvVar block. Dictionaries include built-in support for overriding a key's (variable name) value with another value consistently. User-provided variable values take precedence over generated values, and only a single version of a variable will be present in the resulting EnvVar block.

#### Special notes for your reviewer:

Previously, based on observed behavior in our test environments, we believed that the last instance of a variable in an EnvVar block took precedence in the running container, so user-provided variables were rendered after the generated variables. As seen in https://github.com/helm/charts/issues/18715, Kubernetes' behavior in this scenario is actually undefined: some providers give precedence to the first variable instance in the block. As discussed in https://github.com/kubernetes/kubernetes/pull/59593, providing duplicate variable definitions should not actually be permitted at all. However, per https://github.com/kubernetes/kubernetes/issues/64841, fixing this (i.e. rejecting configurations that include multiple definitions of a variable) is fairly difficult, and Kubernetes will continue to accept multiple instances of a variable going forward.

At present, I believe we should actually reverse our current precedence order: there should not be many cases where it makes sense for users to override generated values (if there are, we should consider those bugs in the value generators and fix them), and overriding values can break the deployment in unexpected ways (e.g. providing your own KONG_ADMIN_LISTEN directly will typically result in the generated admin API Service not matching the listen in the container/blocking admin API access). We could, if necessary, add an additional dictionary that looks for special override values in `env` (e.g. instead of `admin_listen`, `override_admin_listen`) and give it final precedence. However, in the spirit of the original intent of the old template, the new template gives all user variables precedence.

The new template relies on a magic string: if a value contains "valueFrom", it is passed to a different case in the render loop to generate a secretKeyRef block (effectively, dump the string as-is). This is necessary because values in a dictionary are always strings, and cannot be detected by the existing `if eq $valueType "map[string]interface {}"` case. We furthermore cannot handle these like plain variable values, as those are passed to `quote`, which mangles the `valueFrom` block. While this is imperfect, I think it's reasonable, as at present (K8S API version 1.17), valueFrom is the only multi-line block suppported by the EnvVar specification.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

File: charts/kong/templates/_helpers.tpl [hunk 1]
```diff
@@ -354,156 +320,175 @@ Use the Pod security context defined in Values or set the UID by default
...
+{{- range keys $completeEnv }}
+{{- $val := pluck . $completeEnv | first -}}
+{{- $valueType := printf "%T" $val -}}
+{{ if eq $valueType "map[string]interface {}" }}
+- name: {{ . }}
+{{ toYaml $val | indent 2 -}}
+{{- else if eq $valueType "string" }}
+{{- if regexMatch "valueFrom" $val }}
+- name: {{ . }}
+{{ $val | indent 2 }}
```
Comment (hbagdi) [inline charts/kong/templates/_helpers.tpl:474]: Can we indent this set of for loops and ifs for a little bit readability?
There are no conventions but having something in this one would help long-term.

**Review by hbagdi — CHANGES_REQUESTED** [review]
This is amazing work. I tested locally a bit and it is pretty neat.

File: charts/kong/templates/_helpers.tpl [hunk 2]
```diff
@@ -354,156 +320,175 @@ Use the Pod security context defined in Values or set the UID by default
...
+{{- end -}}
+
+{{/*
+      ====== MERGE AND RENDER ENV BLOCK ======
+*/}}
+
+{{- $completeEnv := mergeOverwrite $autoEnv $userEnv -}}
+
+{{- range keys $completeEnv }}
+{{- $val := pluck . $completeEnv | first -}}
```
[the remark this answers is not in the document]
Comment (rainest) [reply charts/kong/templates/_helpers.tpl:466]: The `mergeOverwrite` does overwrite duplicates, and does so deterministically: the value of a duplicate key in the rightmost argument (`$userEnv` here) is used. The order of elements within the resulting dictionary, and therefore in the rendered template, is not deterministic, but that doesn't affect the resulting environment.

The use of `pluck key dict | first` is a Helm 2 compatibility thing: this originally used the simpler `get` function, but that didn't exist until [Sprig v3.0.2](https://github.com/Masterminds/sprig/pull/197). Helm 2 uses Sprig v2.x, and probably will do so forever. `pluck key dict | first` is the older Sprig 2 idiomatic way of working around the lack of a dedicated "get the only value for a key out of dict" function.

File: charts/kong/templates/_helpers.tpl [hunk 3]
```diff
@@ -354,156 +320,175 @@ Use the Pod security context defined in Values or set the UID by default
...
-{{- end }}
+  {{- $_ := set $autoEnv "KONG_PORTAL_GUI_HOST" .Values.portal.ingress.hostname -}}
+  {{- if .Values.portal.ingress.tls }}
+    {{- $_ := set $autoEnv "KONG_PORTAL_GUI_PROTOCOL" "https" -}}
+  {{- else }}
+    {{- $_ := set $autoEnv "KONG_PORTAL_GUI_PROTOCOL" "http" -}}
+  {{- end }}
 {{- end }}
+
 {{- if and (not .Values.env.portal_api_listen) (.Values.enterprise.enabled) (.Values.enterprise.portal.enabled) }}
```
[the remark this answers is not in the document]
Comment (rainest) [reply charts/kong/templates/_helpers.tpl:376]: Correct: this preserves the existing order of the env block generation, which grew organically over time and needs a reorg. Preserving this here is mostly to make the diff a bit easier to read.

After this, I intend to rebase and PR the version at https://github.com/rainest/charts/blob/d1362e747d43ac9f80c155d7e6ef5d4709161e9c/stable/kong/templates/_helpers.tpl, which reorganizes the block a bit and unifies the listeners (sort of--it keeps one of the old ones with a deprecation warning for now).

File: charts/kong/templates/_helpers.tpl (hunk 1 above)
Comment (rainest) [reply charts/kong/templates/_helpers.tpl:474]: Maybe. It's easy to do this for the blocks above because nothing there gets rendered and all whitespace is chomped, but for the render block, both whitespace and indentation matter, and getting them right is tricky. I can do something like this easily:

```
{{- range keys $completeEnv }}
  {{- $val := pluck . $completeEnv | first -}} 
  {{- $valueType := printf "%T" $val -}} 
  {{ if eq $valueType "map[string]interface {}" }}
- name: {{ . }}
{{ toYaml $val | indent 2 -}} 
  {{- else if eq $valueType "string" }}
    {{- if regexMatch "valueFrom" $val }}
- name: {{ . }}
{{ $val | indent 2 }}
    {{- else }}
- name: {{ . }}
  value: {{ $val | quote }}
    {{- end }}
  {{- else }}
- name: {{ . }}
  value: {{ $val | quote }}
  {{- end }}
{{- end -}} 
```
But IMO this is actually less readable than the flat version, since you can't easily tell whether indentation is something we render or if it's to indicate the depth within control flow blocks. With the flat version, any indentation is something that's rendered in the template.

There may be a way to structure this such that we can freely indent both the control flow stuff and rendered content within so that they both reflect the control flow depth, and indentation for rendered content is controlled independently. A bunch of `{{- printf whatever | indent X}}` statements is probably a viable approach.

However, Helm makes this very annoying to debug when you get it wrong: any invalid YAML breaks the entire template process, and produces nothing more something like this very unhelpful error:
```
Error: YAML parse error on kong/templates/migrations.yaml: error converting YAML to JSON: yaml: line 38: could not find expected ':'
```
The line number there is in the rendered invalid YAML, which Helm refuses to show you. Their official guidance is to comment out the section you think you broke, but it's not always obvious which (sometimes you chomped a newline you shouldn't have and broke a section other than the one you're modifying) and sometimes you can't easily comment everything (with nested templates, you have to comment out content in all of them).

So, ultimately, yes, but not yet, and contingent on me or someone else writing a Helm patch that lets it just dump the offending YAML when it fails to convert to JSON.

File: charts/kong/templates/_helpers.tpl [hunk 4]
```diff
@@ -354,156 +320,175 @@ Use the Pod security context defined in Values or set the UID by default
...
-{{- end }}
-{{- end }}
+  {{- $_ := set $autoEnv "KONG_ADMIN_LISTEN" (printf "0.0.0.0:%d" (int64 .Values.admin.containerPort)) -}}
+{{- end -}}
+
 {{- if .Values.admin.ingress.enabled }}
-- name: KONG_ADMIN_API_URI
-  value: {{ include "kong.ingress.serviceUrl" .Values.admin.ingress }}
-{{- end }}
+  {{- $_ := set $autoEnv "KONG_ADMIN_API_URI" (include "kong.ingress.serviceUrl" .Values.admin.ingress) -}}
```
[the remark this answers is not in the document]
Comment (rainest) [reply charts/kong/templates/_helpers.tpl:348]: Similar to the comment below: this is wrong but should be fixed in https://github.com/rainest/charts/blob/d1362e747d43ac9f80c155d7e6ef5d4709161e9c/stable/kong/templates/_helpers.tpl, as this PR preserves the decision tree from the current env generator.

File: charts/kong/templates/_helpers.tpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (hbagdi) [reply charts/kong/templates/_helpers.tpl:466]: I didn't frame my question correctly.
My question is why do even need in the first place? Will `mergeOverwrite` not take care of removing duplicates? If yes, then we can simply iterate over `$completeEnv` and be done with it. I'm sure I'm missing something here.

File: charts/kong/templates/_helpers.tpl (hunk 1 above)
Comment (hbagdi) [reply charts/kong/templates/_helpers.tpl:474]: Templating YAML with Go templates was such a good idea.

Let's keep it as is. This was more about readability but as as you very well summarized, correctness would be hard to guarantee and it's well worth to sacrifice readability here.
Thank you for the detailed explanations, as always!

**Comment by hbagdi** [comment]
Can you fix the conflicts and we will be good to go?

File: charts/kong/templates/_helpers.tpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (rainest) [reply charts/kong/templates/_helpers.tpl:466]: It does, and we are--iterating over over `{{- range keys $completeEnv -}}` gives us the key in `.`, but not its value, which we have to extract separately. There's nothing in http://masterminds.github.io/sprig/dicts.html that provides both.

`pluck` returns an array of all values for a given key, but we just want the naked value. Absent get, piping that through `first` effectively just discards the wrapping array, since we already know it will only have one value.

**Review by hbagdi — APPROVED** [review]

**Comment by hbagdi** [comment]
@rainest Thank you for this fix. Your persistence and hard-work is truly inspiring!
